pallets/pallet-governance/src/lib.rs

```diff
@@ -0,0 +1,446 @@
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Type representing the weight of this pallet
+		type WeightInfo: WeightInfo;
+
+		/// The outer Origin needs to be compatible with this pallet's Origin
+		type RuntimeOrigin: From<RawOrigin>
+			+ From<frame_system::RawOrigin<<Self as frame_system::Config>::AccountId>>;
+
+		/// The overarching call type.
+		type RuntimeCall: Member
+			+ Parameter
+			+ UnfilteredDispatchable<RuntimeOrigin = <Self as Config>::RuntimeOrigin>
+			+ From<frame_system::Call<Self>>
+			+ From<Call<Self>>
+			+ GetDispatchInfo;
+
+		type BlockNumberProvider: BlockNumberProvider<BlockNumber = BlockNumberFor<Self>>;
+
+		type EnsureGovernance: EnsureOrigin<<Self as frame_system::Config>::RuntimeOrigin>;
+
+		#[pallet::constant]
+		type ExpiryPeriod: Get<BlockNumberFor<Self>>;
+
+		#[pallet::constant]
+		type MajorityThreshold: Get<Percent>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The member is not inside the council.
+		Unauthorized,
+		/// The proposal id does not exist.
+		InvalidProposalId,
+		/// The user has voted.
+		AlreadyVoted,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Registered the person who made the proposal.
+		ProposalRegistered { who: T::AccountId, call: <T as Config>::RuntimeCall },
+
+		/// Registered the vote result.
+		VoteCasted { who: T::AccountId, proposal: ProposalId, approve: bool },
+
+		/// Rotated the council members.
+		AuthorityRotated { new_council: BTreeSet<T::AccountId> },
+
+		/// Registered the passed proposal.
+		ProposalPassed {
+			id: ProposalId,
+			call: <T as Config>::RuntimeCall,
+			result: DispatchResultWithPostInfo,
+		},
+
+		/// Registered the rejected proposal.
+		ProposalRejected { id: ProposalId, reason: RejectReason },
+	}
+
+	/// Stores the next Proposal ID should be.
+	#[pallet::storage]
+	pub type NextProposalId<T: Config> = StorageValue<_, ProposalId, ValueQuery>;
+
+	/// Stores the proposal id related the call.
+	#[pallet::storage]
+	#[pallet::getter(fn proposals)]
+	pub type Proposals<T: Config> = StorageMap<_, Blake2_128Concat, ProposalId, EncodedCall>;
+
+	/// Stores the proposal IDs that are set to expire at a specific block number.
+	#[pallet::storage]
+	#[pallet::getter(fn expiry)]
+	pub type Expiry<T: Config> =
+		StorageMap<_, Blake2_128Concat, BlockNumberFor<T>, BTreeSet<ProposalId>, ValueQuery>;
+
+	/// Stores the vote results of the proposal id.
+	#[pallet::storage]
+	#[pallet::getter(fn votes)]
+	pub type Votes<T: Config> =
+		StorageMap<_, Blake2_128Concat, ProposalId, CastedVotes<T::AccountId>, ValueQuery>;
+
+	/// Stores the current council members.
+	#[pallet::storage]
+	#[pallet::getter(fn authorities)]
+	pub type CurrentAuthorities<T: Config> = StorageValue<_, BTreeSet<T::AccountId>, ValueQuery>;
+
+	/// Stores the resolved proposals.
```

Comment:
This needs to be explained better.

Proposals that can be resolved are stored here, until the next block's on_finalize, where the the proposal is accepted (call dispatched) or rejected.